cmd/endpoint.go

```diff
@@ -0,0 +1,145 @@
+/*
+ * Minio Cloud Storage, (C) 2017 Minio, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cmd
+
+import (
+	"fmt"
+	"net/url"
+
+	"github.com/minio/minio-go/pkg/set"
+)
+
+// IsValidDistribution - checks whether given count is a valid distribution for erasure coding.
+func IsValidDistribution(count int) bool {
+	return (count >= 4 && count <= 16 && count%2 == 0)
+}
+
+// IsEmptyPath - check whether given path is not empty.
+func IsEmptyPath(path string) bool {
+	return path == "" || path == "." || path == "/" || path == `\`
+}
+
+// EndpointType - enum for endpoint type.
+type EndpointType int
+
+const (
+	// UnknownEndpointType - unknown endpoint type enum.
+	UnknownEndpointType EndpointType = 0
+
+	// FSEndpointType - FS endpoint type enum.
+	FSEndpointType = 1
+
+	// XLEndpointType - XL endpoint type enum.
+	XLEndpointType = 2
+)
+
+// Endpoint - any type of endpoint.
+type Endpoint struct {
+	Index      int
+	Value      string
+	IsDisabled bool
+	URL        *url.URL
+}
+
+// Type - returns type of endpoint.
+func (endpoint Endpoint) Type() EndpointType {
+	if IsEmptyPath(endpoint.Value) {
+		return UnknownEndpointType
+	}
+
+	if endpoint.URL != nil {
+		return XLEndpointType
+	}
+
+	return FSEndpointType
+}
+
+// NewEndpoint - returns new endpoint based on given arguments.
+func NewEndpoint(arg string, index int) (Endpoint, error) {
+	if IsEmptyPath(arg) {
+		return Endpoint{}, fmt.Errorf("Empty or root endpoint is not supported")
+	}
+
+	u, err := url.Parse(arg)
+	if err == nil && u.Host != "" {
+		// May be XL type endpoint arg.
+		host, port, err := splitHostPort(u.Host)
+		if !(err == nil && host != "" && (u.Scheme == "http" || u.Scheme == "https") &&
```

Comment:
At this point, I think we can check for Scheme, Opaque, ForceQuery, RawQuery and Fragment and also IsEmptyPath() just before calling splitHostPost(). It will be more simple to read.